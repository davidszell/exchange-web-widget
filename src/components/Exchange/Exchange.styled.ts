import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding: 50px;
  border: 2px solid #157ae2;
  border-radius: 10px;
  width: 500px;
  color: #222;
  gap: 20px;
`;

export const Heading = styled.div`
  display: flex;
  flex-direction: column;
  gap: 5px;
`;

export const Header = styled.h1`
  margin: 0;
  color: #222;
`;

export const SubHeader = styled.p`
  margin: 0;
  color: #777;
  font-size: 0.8em;
`;

export const Button = styled.button`
  border: 1px solid #157ae2;
  color: #157ae2;
  background-color: white;
  padding: 10px;
  border-radius: 5px;
  cursor: pointer;
  font-weight: 500;
  &:hover {
    color: white;
    background-color: #157ae2;
  }
`;

export const ExchangeButton = styled(Button)`
  width: 100%;
`;

export const DirectionButton = styled(Button)`
  border-color: #777;
  color: #777;
  font-size: 1.2em;

  &:hover {
    border-color: #157ae2;
    color: #157ae2;
    background-color: white;
  }
`;
