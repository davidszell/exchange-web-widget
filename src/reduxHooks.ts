import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { RootState } from "./store";

export const useAppDispatch = (): ThunkDispatch<RootState, any, AnyAction> => useDispatch<ThunkDispatch<RootState, any, AnyAction>>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector