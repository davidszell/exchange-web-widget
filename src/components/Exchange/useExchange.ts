import { useEffect, useMemo, useState } from 'react';
import { doExchange, fetchExchangeRates } from '../../actions';
import { useAppDispatch, useAppSelector } from '../../reduxHooks';
import { ExchangeRatesType, WalletType } from '../../types';

export type ExchangeDirectionType = 'buy' | 'sell';

const useExchange = () => {
  const wallets: WalletType[] = useAppSelector((state) => state.wallets);
  const exchangeRates: ExchangeRatesType = useAppSelector((state) => state.exchangeRates);

  const [fromWallet, setFromWallet] = useState<WalletType>(wallets[0]);
  const [toWallet, setToWallet] = useState<WalletType>(wallets[1]);
  const [fromAmount, setFromAmount] = useState<number | null>(null);
  const [toAmount, setToAmount] = useState<number | null>(null);
  const [direction, setDirection] = useState<ExchangeDirectionType>('sell');
  const [disableExchange, setDisableExchange] = useState<boolean>(true);

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchExchangeRates());

    const interval = setInterval(() => {
      dispatch(fetchExchangeRates());
    }, 10000);

    return () => clearInterval(interval);
  }, [dispatch]);

  useEffect(() => {
    setFromWallet((prevFromWallet) => wallets.find((wallet) => wallet.name === prevFromWallet.name) || wallets[0]);
    setToWallet((prevToWallet) => wallets.find((wallet) => wallet.name === prevToWallet.name) || wallets[1]);
  }, [wallets]);

  useEffect(() => {
    setDisableExchange(toAmount === null
      || fromAmount === null
      || fromAmount === 0
      || toAmount === 0
      || (direction === 'sell' && fromAmount > fromWallet.balance)
      || (direction === 'buy' && toAmount > toWallet.balance));
  }, [direction, fromAmount, fromWallet.balance, toAmount, toWallet.balance]);

  const roundRates = (value: number) => Number(`${Math.round(Number(`${value}e2`))}e-2`);

  const exchangeRate: number = useMemo(() => {
    if (fromWallet.name === exchangeRates.base) {
      return exchangeRates.rates[toWallet.name];
    }
    const fromRate = exchangeRates.rates[fromWallet.name];
    const toRate = exchangeRates.rates[toWallet.name];

    return toRate / fromRate;
  }, [fromWallet, toWallet, exchangeRates]);

  const handleFromWalletChange = (walletName: string) => {
    if (fromWallet.name === walletName) {
      return;
    }

    const newCurrency = wallets.find((item) => item.name === walletName);

    if (!newCurrency) {
      return;
    }

    if (toWallet.name === walletName) {
      setToWallet(fromWallet);
    }

    setFromWallet(newCurrency);
    setFromAmount(0);
    setToAmount(0);
  };

  const handleToWalletChange = (walletName: string) => {
    if (toWallet.name === walletName) {
      return;
    }

    const newCurrency = wallets.find((item) => item.name === walletName);

    if (!newCurrency) {
      return;
    }

    if (fromWallet.name === walletName) {
      setFromWallet(toWallet);
    }

    setToWallet(newCurrency);
    setFromAmount(0);
    setToAmount(0);
  };

  const handleFromAmountChange = (newAmount: number | null) => {
    if (newAmount !== null && newAmount > 0) {
      setFromAmount(newAmount);
      setToAmount(roundRates(newAmount * exchangeRate));
    } else {
      setToAmount(null);
      setFromAmount(null);
    }
  };

  const handleToAmountChange = (newAmount: number) => {
    if (newAmount !== null && newAmount > 0) {
      setToAmount(newAmount);
      setFromAmount(roundRates(newAmount / exchangeRate));
    } else {
      setToAmount(null);
      setFromAmount(null);
    }
  };

  const toggleDirection = () => {
    setDirection((prevDirection) => (prevDirection === 'sell' ? 'buy' : 'sell'));
  };

  const handleExchange = () => {
    dispatch(doExchange(fromWallet.name, toWallet.name, direction, fromAmount, toAmount));
    setFromAmount(0);
    setToAmount(0);
  };

  return {
    fromWallet,
    toWallet,
    fromAmount,
    toAmount,
    exchangeRate,
    direction,
    disableExchange,
    handleFromWalletChange,
    handleToWalletChange,
    handleFromAmountChange,
    handleToAmountChange,
    handleExchange,
    toggleDirection,
  };
};

export default useExchange;
