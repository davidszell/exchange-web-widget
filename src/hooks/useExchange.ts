import { useEffect, useState } from 'react';
import { doExchange, fetchExchangeRates } from '../actions';
import { useAppDispatch, useAppSelector } from './reduxHooks';
import { ExchangeRatesType, WalletType } from '../types';
import { formatFloatRound } from '../helpers';

export type ExchangeActionType = 'buy' | 'sell';

const useExchange = () => {
  const wallets: WalletType[] = useAppSelector((state) => state.wallets);
  const exchangeRates: ExchangeRatesType = useAppSelector((state) => state.exchangeRates);

  const [sourceWallet, setSourceWallet] = useState<WalletType>(wallets[0]);
  const [destinationWallet, setDestinationWallet] = useState<WalletType>(wallets[1]);
  const [sourceAmount, setSourceAmount] = useState<number | null>(null);
  const [destinationAmount, setDestinationAmount] = useState<number | null>(null);
  const [exchangeAction, setExchangeAction] = useState<ExchangeActionType>('sell');
  const [isExchangeDisabled, setExchangeDisabled] = useState<boolean>(true);
  const [exchangeRate, setExchangeRate] = useState<number | null>(null);

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchExchangeRates());

    const interval = setInterval(() => {
      dispatch(fetchExchangeRates());
    }, 10000);

    return () => clearInterval(interval);
  }, [dispatch]);

  useEffect(() => {
    setSourceWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[0]);
    setDestinationWallet((prevWallet) => wallets.find((wallet) => wallet.name === prevWallet.name) || wallets[1]);
  }, [wallets]);

  useEffect(() => {
    setExchangeDisabled(
      sourceAmount === null
      || destinationAmount === null
      || (exchangeAction === 'sell' && sourceAmount > sourceWallet.balance)
      || (exchangeAction === 'buy' && destinationAmount > destinationWallet.balance),
    );
  }, [exchangeAction, sourceAmount, sourceWallet.balance, destinationAmount, destinationWallet.balance]);

  useEffect(() => {
    if (Object.keys(exchangeRates.rates).length === 0) {
      setExchangeRate(null);
      return;
    }

    const sourceRate = exchangeRates.rates[sourceWallet.name];
    const destinationRate = exchangeRates.rates[destinationWallet.name];

    if (sourceRate == null || destinationRate == null) {
      setExchangeRate(null);
      return;
    }

    if (sourceWallet.name === exchangeRates.base) {
      setExchangeRate(exchangeRates.rates[destinationWallet.name]);
    } else {
      setExchangeRate(destinationRate / sourceRate);
    }
  }, [sourceWallet, destinationWallet, exchangeRates]);

  const clearAmounts = () => {
    setSourceAmount(null);
    setDestinationAmount(null);
  };

  const handleSourceWalletChange = (walletName: string) => {
    if (sourceWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (destinationWallet.name === walletName) {
      setDestinationWallet(sourceWallet);
    }

    setSourceWallet(newWallet);
    clearAmounts();
  };

  const handleDestinationWalletChange = (walletName: string) => {
    if (destinationWallet.name === walletName) {
      return;
    }

    const newWallet = wallets.find((item) => item.name === walletName);

    if (!newWallet) {
      return;
    }

    if (sourceWallet.name === walletName) {
      setSourceWallet(destinationWallet);
    }

    setDestinationWallet(newWallet);
    clearAmounts();
  };

  const handleSourceAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setSourceAmount(newAmount);
      setDestinationAmount(formatFloatRound(newAmount * exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const handleDestinationAmountChange = (newAmount: number | null) => {
    if (exchangeRate !== null && newAmount !== null && newAmount > 0) {
      setDestinationAmount(newAmount);
      setSourceAmount(formatFloatRound(newAmount / exchangeRate));
    } else {
      clearAmounts();
    }
  };

  const toggleExchangeAction = () => {
    setExchangeAction((prevExchangeAction) => (prevExchangeAction === 'sell' ? 'buy' : 'sell'));
  };

  const handleExchange = () => {
    if (!isExchangeDisabled) {
      dispatch(doExchange(sourceWallet.name, destinationWallet.name, exchangeAction, sourceAmount, destinationAmount));
      clearAmounts();
    }
  };

  return {
    exchangeAction,
    exchangeRate,
    isExchangeDisabled,
    sourceWallet,
    destinationWallet,
    sourceAmount,
    destinationAmount,
    handleExchange,
    handleSourceWalletChange,
    handleDestinationWalletChange,
    handleSourceAmountChange,
    handleDestinationAmountChange,
    toggleExchangeAction,
  };
};

export default useExchange;
