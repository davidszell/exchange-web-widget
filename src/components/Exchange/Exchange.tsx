import React from 'react';
import ExchangeItem from '../ExchangeItem';
import {
  Container, DirectionButton, ExchangeButton, Header, Heading, SubHeader,
} from './Exchange.styled';
import useExchange from './useExchange';

const Exchange = (): JSX.Element => {
  const {
    fromWallet,
    toWallet,
    fromAmount,
    toAmount,
    exchangeRate,
    direction,
    handleFromWalletChange,
    handleToWalletChange,
    handleFromAmountChange,
    handleToAmountChange,
    handleExchange,
    toggleDirection,
  } = useExchange();

  const handleExchangeDirectionChange = () => {
    toggleDirection();
  };

  return (
    <Container>
      <Heading>
        <Header data-testid="header">
          {direction === 'sell' ? 'Sell' : 'Buy'}
          {' '}
          {fromWallet.name}
        </Header>
        <SubHeader data-testid="exchangeRate">
          {fromWallet.symbol}
          1 =
          {' '}
          {toWallet.symbol}
          {exchangeRate}
        </SubHeader>
      </Heading>
      <ExchangeItem
        wallet={fromWallet}
        handleWalletChange={handleFromWalletChange}
        amount={fromAmount}
        handleAmountChange={handleFromAmountChange}
      />
      <DirectionButton data-testid="directionButton" onClick={handleExchangeDirectionChange}>{direction === 'sell' ? '↓' : '↑'}</DirectionButton>
      <ExchangeItem
        wallet={toWallet}
        handleWalletChange={handleToWalletChange}
        amount={toAmount}
        handleAmountChange={handleToAmountChange}
      />
      <ExchangeButton data-testid="exchangeButton" onClick={handleExchange}>
        {direction === 'sell' ? 'Sell' : 'Buy'}
        {' '}
        {fromWallet.name}
        {' '}
        {direction === 'sell' ? 'for' : 'with'}
        {' '}
        {toWallet.name}
      </ExchangeButton>
    </Container>
  );
};

export default Exchange;
