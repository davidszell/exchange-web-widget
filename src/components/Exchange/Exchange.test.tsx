import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import Exchange from './Exchange';
import * as reduxHooks from '../../hooks/reduxHooks';
import * as useExchange from '../../hooks/useExchange';

jest.mock('../../hooks/useExchange', () => {
  const original = jest.requireActual('../../hooks/useExchange')
  return {
    __esModule: true,
    default: jest.fn(() => ({
      ...original.default(),
      exchangeRate: 0.85
    }))
  }
});

describe('Exchange', () => {

  const useSelectorMock = jest.spyOn(reduxHooks, 'useAppSelector')
  const useDispatchMock = jest.spyOn(reduxHooks, 'useAppDispatch')

  beforeEach(() => {
    useSelectorMock.mockImplementation(selector => selector(mockStore));
    useDispatchMock.mockReturnValue(jest.fn())
  })
  afterEach(() => {
    useSelectorMock.mockClear()
    useDispatchMock.mockClear()
  })

  const mockStore = {
    wallets: [
      {
        name: "EUR",
        balance: 10,
        longName: "Euros",
        symbol: "€"
      },
      {
        name: "GBP",
        balance: 20,
        longName: "British Pounds",
        symbol: "£"
      },
      {
        name: "USD",
        balance: 30,
        longName: "US Dollars",
        symbol: "$"
      }
    ],
    exchangeRates: {
      base: "EUR",
      rates: {
        "EUR": 1,
        "GBP": 0.85,
        "USD": 1.16
      }
    }
  };

  it('matches snapshot', () => {
    const tree = renderer
      .create(<Exchange />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shows correct header', () => {
    render(<Exchange />);
    expect(screen.getByTestId('header')).toHaveTextContent(/^Sell [A-Z]{3}$/);
  });

  it('Shows correct header after exchange action change', () => {
    render(<Exchange />);
    fireEvent.click(screen.getByTestId('directionButton'));
    expect(screen.getByTestId('header')).toHaveTextContent(/^Buy [A-Z]{3}$/);
  });

  it('Shows correct button label', () => {
    render(<Exchange />);
    expect(screen.getByTestId('directionButton')).toHaveTextContent(/^↓$/);
  });

  it('Shows correct button after exchange action change', () => {
    render(<Exchange />);
    fireEvent.click(screen.getByTestId('directionButton'));
    expect(screen.getByTestId('directionButton')).toHaveTextContent(/^↑$/);
  });

  it('Shows correct exchange button label', () => {
    render(<Exchange />);
    expect(screen.getByTestId('exchangeButton')).toHaveTextContent(/^Sell [A-Z]{3} for [A-Z]{3}$/);
  });

  it('Shows correct exchange button label after exchange action change', () => {
    render(<Exchange />);
    fireEvent.click(screen.getByTestId('directionButton'));
    expect(screen.getByTestId('exchangeButton')).toHaveTextContent(/^Buy [A-Z]{3} with [A-Z]{3}$/);
  });

  it('Shows loading screen when echangeRate is null', () => {
    (useExchange.default as jest.Mock).mockImplementation(() => ({
      exchangeRate: null
    }))
    render(<Exchange />);
    expect(screen.queryByText('Loading...')).not.toBeNull();
  })
})