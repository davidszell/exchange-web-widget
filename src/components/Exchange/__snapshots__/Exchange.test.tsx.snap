// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange matches snapshot 1`] = `
.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: none;
  text-align: right;
  font-size: 1.3em;
  padding: 10px;
  border-radius: 5px;
  -moz-appearance: textfield;
}

.c8::-webkit-outer-spin-button,
.c8::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 50px;
  border: 2px solid #157ae2;
  border-radius: 10px;
  width: 500px;
  color: #222;
  gap: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 5px;
}

.c2 {
  margin: 0;
  color: #222;
}

.c3 {
  margin: 0;
  color: #777;
  font-size: 0.8em;
}

.c6 {
  border: 1px solid #157ae2;
  color: #157ae2;
  background-color: white;
  padding: 10px;
  border-radius: 5px;
  cursor: pointer;
  font-weight: 500;
}

.c6:hover {
  color: white;
  background-color: #157ae2;
}

.c6:disabled,
.c6[disabled],
.c6:disabled:hover,
.c6[disabled]:hover {
  cursor: unset;
  color: #777;
  border-color: #777;
  background-color: white;
}

.c11 {
  width: 100%;
}

.c10 {
  border-color: #777;
  color: #777;
  font-size: 1.2em;
}

.c10:hover {
  border-color: #157ae2;
  color: #157ae2;
  background-color: white;
}

.c4 {
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 10px;
}

.c7::after {
  content: '⮁';
  margin-left: 10px;
}

.c9 {
  font-size: 0.8em;
  margin: 0;
  color: #777;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <h1
      className="c2"
      data-testid="header"
    >
      Sell
       
      EUR
    </h1>
    <p
      className="c3"
      data-testid="exchangeRate"
    >
      €
      1 =
       
      £
      0.85
    </p>
  </div>
  <div
    className="c4"
  >
    <div
      className="c5"
    >
      <button
        className="c6 c7"
        onClick={[Function]}
      >
        EUR
      </button>
      <input
        className="c8"
        onChange={[Function]}
        placeholder="0"
        step="0.01"
        type="number"
        value=""
      />
    </div>
    <p
      className="c9"
    >
      Balance:
      €
      10
    </p>
    
  </div>
  <button
    className="c6 c10"
    data-testid="directionButton"
    onClick={[Function]}
  >
    ↓
  </button>
  <div
    className="c4"
  >
    <div
      className="c5"
    >
      <button
        className="c6 c7"
        onClick={[Function]}
      >
        GBP
      </button>
      <input
        className="c8"
        onChange={[Function]}
        placeholder="0"
        step="0.01"
        type="number"
        value=""
      />
    </div>
    <p
      className="c9"
    >
      Balance:
      £
      20
    </p>
    
  </div>
  <button
    className="c6 c11"
    data-testid="exchangeButton"
    disabled={true}
    onClick={[Function]}
  >
    Sell
     
    EUR
     
    for
     
    GBP
  </button>
</div>
`;
